cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageTest.java

```diff
@@ -77,13 +87,120 @@ public void testLineage() throws Exception {
+    Id.Program workflow = Id.Program.from(app, ProgramType.WORKFLOW, AllProgramsApp.NoOpWorkflow.NAME);
+    Id.DatasetInstance dataset = Id.DatasetInstance.from(namespace, AllProgramsApp.DATASET_NAME);
+
+    Assert.assertEquals(200, status(createNamespace(namespace)));
+    Assert.assertEquals(200,
+                        status(deploy(AllProgramsApp.class, Constants.Gateway.API_VERSION_3_TOKEN, namespace)));
+    try {
+      addProperties(dataset, ImmutableMap.of("data-key1", "data-value1"));
+      Assert.assertEquals(ImmutableMap.of("data-key1", "data-value1"), getProperties(dataset));
+
+      // Start all programs
+      RunId flowRunId = runAndWait(flow);
+      RunId mrRunId = runAndWait(mapreduce);
+      RunId sparkRunId = runAndWait(spark);
+      runAndWait(workflow);
+      RunId workflowMrRunId = getRunId(mapreduce, mrRunId);
+      RunId serviceRunId = runAndWait(service);
+      // Worker makes a call to service to make it access datasets,
+      // hence need to make sure service starts before worker, and stops after it.
+      RunId workerRunId = runAndWait(worker);
+
+      // Wait for programs to finish
+      waitForStop(flow, true);
+      waitForStop(mapreduce, false);
+      waitForStop(spark, false);
+      waitForStop(workflow, false);
+      waitForStop(worker, false);
+      waitForStop(service, true);
+
+      long now = System.currentTimeMillis();
+      long oneHourMillis = TimeUnit.HOURS.toMillis(1);
+      HttpResponse httpResponse = fetchLineage(dataset, now - oneHourMillis, now + oneHourMillis, 10);
+      LineageRecord lineage = GSON.fromJson(httpResponse.getResponseBodyAsString(), LineageRecord.class);
+
+      // dataset is accessed by all programs
+      LineageRecord expected =
+        new LineageRecord(now - oneHourMillis, now + oneHourMillis,
+                          ImmutableSet.of(
+                            new Relation(dataset, flow, AccessType.UNKNOWN, ImmutableSet.of(flowRunId),
+                                         ImmutableSet.of(Id.Flow.Flowlet.from(flow, AllProgramsApp.A.NAME))),
+                            new Relation(dataset, mapreduce, AccessType.UNKNOWN, ImmutableSet.of(mrRunId)),
+                            new Relation(dataset, spark, AccessType.UNKNOWN, ImmutableSet.of(sparkRunId)),
+                            new Relation(dataset, mapreduce, AccessType.UNKNOWN, ImmutableSet.of(workflowMrRunId)),
+                            new Relation(dataset, service, AccessType.UNKNOWN, ImmutableSet.of(serviceRunId)),
+                            new Relation(dataset, worker, AccessType.UNKNOWN, ImmutableSet.of(workerRunId))
+                          ));
+      Assert.assertEquals(expected, lineage);
+    } finally {
+      try {
+        deleteApp(app, 200);
+      } catch (Throwable e) {
+        LOG.error("Got exception while deleting app", e);
+      }
+    }
+  }
+
+  private RunId runAndWait(Id.Program program) throws Exception {
```

Comment:
all the methods from here on seem pretty useful, should we move them to ``AppFabricTestBase``?